src/p2p/pubsub.rs

```diff
@@ -0,0 +1,352 @@
+        &mut self,
+        peer_id: PeerId,
+        event: <Self::ProtocolsHandler as ProtocolsHandler>::OutEvent,
+    ) {
+        self.floodsub.inject_node_event(peer_id, event)
+    }
+
+    fn inject_addr_reach_failure(
+        &mut self,
+        peer_id: Option<&PeerId>,
+        addr: &Multiaddr,
+        error: &dyn std::error::Error,
+    ) {
+        self.floodsub
+            .inject_addr_reach_failure(peer_id, addr, error)
+    }
+
+    fn poll(
+        &mut self,
+        ctx: &mut Context,
+        poll: &mut impl PollParameters,
+    ) -> Poll<PubsubNetworkBehaviourAction> {
+        use futures::stream::StreamExt;
+        use std::collections::hash_map::Entry;
+
+        loop {
+            match self.unsubscriptions.1.poll_next_unpin(ctx) {
+                Poll::Ready(Some(dropped)) => {
+                    let topic = Topic::new(dropped);
+                    if self.streams.remove(&topic).is_some() {
+                        log::debug!("unsubscribing via drop from {:?}", topic.id());
+                        assert!(
+                            self.floodsub.unsubscribe(topic),
+                            "Failed to unsubscribe a dropped subscription"
+                        );
+                    } else {
+                        // unsubscribed already by `unsubscribe`
+                        // TODO: not sure if the unsubscribe functionality is needed as the
+                        // unsubscribe on drop seems to work
+                    }
+                }
+                Poll::Ready(None) => unreachable!("we own the sender"),
+                Poll::Pending => break,
+            }
+        }
+
+        match futures::ready!(self.floodsub.poll(ctx, poll)) {
+            NetworkBehaviourAction::GenerateEvent(FloodsubEvent::Message(msg)) => {
+                let topics = msg.topics.clone();
+                let msg = Arc::new(PubsubMessage::from(msg));
+                let mut buffer = None;
+
+                for topic in topics {
+                    match self.streams.entry(topic) {
+                        Entry::Occupied(oe) => {
+                            let sent = buffer.take().unwrap_or_else(|| Arc::clone(&msg));
+                            if let Err(se) = oe.get().unbounded_send(sent) {
+                                // receiver has dropped
+                                let (topic, _) = oe.remove_entry();
+                                log::debug!("unsubscribing via SendError from {:?}", topic.id());
+                                assert!(
+                                    self.floodsub.unsubscribe(topic.clone()),
+                                    "Closed sender didnt allow unsubscribing {:?}",
+                                    topic
+                                );
+                                buffer = Some(se.into_inner());
+                            }
+                        }
+                        Entry::Vacant(ve) => {
+                            panic!(
```

Comment:
does this mean I can crash the ipfs node by sending it a message tagged with a topic it's not subscribed to?

Reply:
The idea was that `Floodsub` will only forward messages to topics we both (`self` in this line, and `floodsub`) agree on that are subscribed. I was hoping to run into this panic while testing but haven't yet.

Reply:
Tried to make a test for this but that is not the way to go. I think the only time this can happen is:

 1. we have been previously subscribing to this topic
 2. we have received a message to the topic
 3. unsubscribed so late than floodsub received the event so that floodsub has already pushed the message on to it's `Floodsub::events`
 4. panic

I'll remove this branch and make this a `if let Entry::Occupied = ...`.

Reply:
You can log it if you prefer